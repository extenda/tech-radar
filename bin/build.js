const fs = require('fs-extra')
const klawSync = require('klaw-sync')
const path = require('path')
const yaml = require('yaml')
const handlebars = require('handlebars')
const md = require('markdown-it')()
const radarModel = require('./radar_model')

const outputDir = 'build'

// Build the static radar HTML representation from the model and radar entries
function buildRadar(archive) {

  handlebars.registerPartial('footer', handlebars.compile(fs.readFileSync('templates/footer.hbs', 'utf8')))
  handlebars.registerHelper('markdown', value => md.render(value))

  const entryPage = handlebars.compile(fs.readFileSync('templates/entry_page.hbs', 'utf8'))
  const quadrantPage = handlebars.compile(fs.readFileSync('templates/quadrant_page.hbs', 'utf8'))
  const radarPage = handlebars.compile(fs.readFileSync('templates/index.hbs', 'utf8'))

  console.log(`Clean ${outputDir}`)
  fs.removeSync(outputDir)
  fs.mkdirSync(outputDir)
  fs.mkdirSync(path.join(outputDir, 'entries'))

  console.log('Build radar')

  var radar = radarModel.createModel()
  radar.date = dateToday()

  console.log('  - Entries')

  // Create all the entries pages and collect blips.
  for (var f of klawSync('radar', {Â nodir: true, traverseAll: true, filter: p => path.extname(p.path) === '.yaml' })) {
    var content = yaml.parse(fs.readFileSync(f.path, 'utf8'))
    validateYaml(f.path, content)
    var quadrantIndex = 0
    findQuadrant(radar, f.path, (i, q) => {
      content.quadrant = q
      quadrantIndex = i
    })

    content.title = radar.title
    content.edition = radar.edition
    content.date = dateToday()
    content.blip.ringName = content.blip.ring.charAt(0) + content.blip.ring.substr(1).toLowerCase()

    const filename = path.join(outputDir, 'entries', path.basename(f.path, '.yaml') + '.html')

    radar.quadrants[quadrantIndex][content.blip.ring.toLowerCase()].push({
      name: content.name,
      file: path.relative(outputDir, filename)
    })

    const blip = createBlip(radar, content, filename, f.path)
    if (blip.active) {
      radar.blips.push(blip)
    }
    fs.writeFileSync(
      filename,
      entryPage(content)
    )
  }

  // Create the quadrant index pages.
  console.log('  - Quadrants')
  for (var q of radar.quadrants) {
    q.title = radar.title
    q.edition = radar.edition
    q.date = dateToday()

    const sortFn = (a, b) => a.name.localeCompare(b.name)
    q.adopt.sort(sortFn)
    q.trial.sort(sortFn)
    q.assess.sort(sortFn)
    q.hold.sort(sortFn)

    fs.writeFileSync(path.join(outputDir, q.dirname + '.html'), quadrantPage(q))
  }

  // Create the main radar visualization
  console.log('  - Radar visualization')
  fs.writeFileSync(path.join(outputDir, 'index.html'), radarPage(radar))

  // Archive the current edition and write the archive index
  createArchive(radar, archive)

  // Copy static resources
  console.log('Copy resources')
  fs.copySync('app', outputDir)
}

function dateToday() {
  return new Date().toLocaleString('en', {
    year: 'numeric',
    month: 'short',
    day: 'numeric'
  })
}

function createArchive(radar, archive) {

  // Archive current edition
  const archiveDir = path.join('archive', radar.edition)
  if (fs.existsSync(archiveDir)) {
    fs.removeSync(archiveDir)
  }

  fs.mkdirSync(path.join(outputDir, 'archive'))
  fs.mkdirSync(archiveDir)

  if (archive) {
    console.log('Archive')
    fs.copySync(outputDir, archiveDir)
  }

  // Create archive page
  const archivePage = handlebars.compile(fs.readFileSync('templates/archive_page.hbs', 'utf8'))
  var editions = []
  for (var f of klawSync('archive', {nofile: true, depthLimit: 0, filter: p => path.basename(p.path) != radar.edition })) {
    const edition = path.basename(f.path)
    fs.copySync(f.path, path.join(outputDir, 'archive', edition))
    editions.push({
      edition: edition,
      created: f.stats.ctime.toLocaleString('en', {
        year: 'numeric',
        month: 'short',
        day: 'numeric'
      })
    })
  }
  fs.writeFileSync(path.join(outputDir, 'archive', 'index.html'), archivePage({
    title: radar.title,
    name: 'Archive',
    date: dateToday(),
    edition: radar.edition,
    editions: editions
  }))
}

function findQuadrant(radar, sourceFile, callback) {
  const quadrant = path.relative('radar', path.dirname(sourceFile))
  for (const [i, q] of radar.quadrants.entries()) {
    if (q.dirname === quadrant) {
      callback(i, q)
      break
    }
  }
}

// Validate all required properties in the radar YAML entries
function validateYaml(filename, content) {
  if (!content.hasOwnProperty('name')) {
    console.error(`Missing 'name' in ${filename}`)
  }
  if (!content.hasOwnProperty('blip')) {
    console.error(`Missing 'blip' in ${filename}`)
  }
  if (!content.blip.hasOwnProperty('since')) {
    console.error(`Missing 'blip.since' in ${filename}`)
  }
  if (!content.blip.hasOwnProperty('ring')) {
    console.error(`Missing 'blip.ring' in ${filename}`)
  }
  if (!content.hasOwnProperty('description')) {
    console.error(`Missing 'description' in ${filename}`)
  }
}

function createBlip(radar, entry, htmlFile, sourceFile) {
  var blip = {
    label: entry.shortname ? entry.shortname : entry.name,
    quadrant: 0,
    ring: radar.rings.indexOf(entry.blip.ring),
    moved: entry.blip.since === radar.edition,
    link: path.relative(outputDir, htmlFile),
    active: true
  }

  if (!blip.moved && entry.blip.hasOwnProperty('moved')) {
    blip.moved = entry.blip.moved
  }

  findQuadrant(radar, sourceFile, (i, q) => blip.quadrant = i)
  if (entry.blip.hasOwnProperty('active')) {
    blip.active = entry.blip.active
  }

  return blip
}

module.exports = {
  build: buildRadar
}

// Command-line entry point.
process.argv.slice(2).forEach(function(val, index, array) {
  if (val === '--build') {
    buildRadar(true)
  }
})
