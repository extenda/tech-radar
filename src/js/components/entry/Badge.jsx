import React from 'react';
import PropTypes from 'prop-types';
import Icon from '../Icon';

const Badge = (props) => {
  const { className, icon, text } = props;

  return (
    <span className={`badge ${className}`}>
      <Icon name={icon} spaceBefore={false} />
      {text}
    </span>
  );
};

Badge.propTypes = {
  className: PropTypes.string,
  icon: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
};

Badge.defaultProps = {
  className: '',
};

export default Badge;
