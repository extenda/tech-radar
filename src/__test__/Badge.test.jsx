import React from 'react';
import { shallow } from 'enzyme/build';
import Badge from '../js/components/entry/Badge';

describe('<Badge />', () => {
  test('It renders badge with icon', () => {
    const component = shallow(<Badge icon="test" text="test" />);
    expect(component).toMatchSnapshot();
  });

  test('It renders badge with extra className', () => {
    const component = shallow(<Badge icon="test" text="test" className="my-test" />);
    expect(component).toMatchSnapshot();
  });
});
