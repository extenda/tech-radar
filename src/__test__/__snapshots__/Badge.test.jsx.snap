// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> It renders badge with extra className 1`] = `
<span
  className="badge my-test"
>
  <Icon
    name="test"
    spaceBefore={false}
  />
  test
</span>
`;

exports[`<Badge /> It renders badge with icon 1`] = `
<span
  className="badge "
>
  <Icon
    name="test"
    spaceBefore={false}
  />
  test
</span>
`;
